packages/parameters/l1parameters.go

```
@@ -2,63 +2,64 @@ package parameters
+	b, err := bcs.Marshal(l)
+	if err != nil {
+		panic(err)
+	}
+	return b
+}
+
+func (l *L1Params) Hash() hashing.HashValue {
+	res, _ := hashing.HashValueFromBytes(l.Bytes())
+	return res
 }
 
 func (l *L1Params) Clone() *L1Params {
 	protocol := l.Protocol.Clone()
 	return &L1Params{
-		MaxPayloadSize: l.MaxPayloadSize,
-		Protocol:       &protocol,
-		BaseToken:      l.BaseToken,
+		Protocol:  &protocol,
+		BaseToken: l.BaseToken,
 	}
 }
 
 type Protocol struct {
-	Epoch                 *iotajsonrpc.BigInt
-	ProtocolVersion       *iotajsonrpc.BigInt
-	SystemStateVersion    *iotajsonrpc.BigInt
-	IotaTotalSupply       *iotajsonrpc.BigInt
-	ReferenceGasPrice     *iotajsonrpc.BigInt
-	EpochStartTimestampMs *iotajsonrpc.BigInt
-	EpochDurationMs       *iotajsonrpc.BigInt
+	Epoch                 *iotajsonrpc.BigInt `json:"epoch" swagger:"required"`
+	ProtocolVersion       *iotajsonrpc.BigInt `json:"protocol_version" swagger:"required"`
+	SystemStateVersion    *iotajsonrpc.BigInt `json:"system_state_version" swagger:"required"`
+	IotaTotalSupply       *iotajsonrpc.BigInt `json:"iota_total_supply" swagger:"required"`
+	ReferenceGasPrice     *iotajsonrpc.BigInt `json:"reference_gas_price" swagger:"required"`
+	EpochStartTimestampMs *iotajsonrpc.BigInt `json:"epoch_start_timestamp_ms" swagger:"required"`
+	EpochDurationMs       *iotajsonrpc.BigInt `json:"epoch_duration_ms" swagger:"required"`
 }
 
 func (p *Protocol) String() string {
-	ret := "{\n"
-	ret += fmt.Sprintf("\tEpoch: %s\n", p.Epoch)
-	ret += fmt.Sprintf("\tProtocolVersion: %s\n", p.ProtocolVersion)
-	ret += fmt.Sprintf("\tSystemStateVersion: %s\n", p.SystemStateVersion)
-	ret += fmt.Sprintf("\tIotaTotalSupply: %s\n", p.IotaTotalSupply)
-	ret += fmt.Sprintf("\tReferenceGasPrice: %s\n", p.ReferenceGasPrice)
-	ret += fmt.Sprintf("\tEpochStartTimestampMs: %s\n", p.EpochStartTimestampMs)
-	ret += fmt.Sprintf("\tEpochDurationMs: %s\n", p.EpochDurationMs)
-	ret += "}\n"
-	return ret
+	b, _ := json.Marshal(p)
```

Use `lo.Must()` or `panic()` instead of ignoring errors.

Also, if this is intended as a human-readable string representation, maybe `json.MarshalIndent` is better.

---

done